### PR TITLE
Draft: Implement `Resource` for `()`

### PR DESCRIPTION
**Pull request by Lubba-64**
# Objective
 - Fixes https://github.com/bevyengine/bevy/issues/14640
## Solution
 - Implement resource for () and insert it as a constant resource.
## Testing
 - Wrote a test to make sure all worlds spawn with () available as a resource.

### CODE REVIEW

**Comment by Lubba-64** [comment]
Giving this another shot. I expect this not to work immediately but I have some more ideas about how to overcome the `#[reflect(Resource)]` problem

**Comment by Lubba-64** [comment]
I believe
```
impl GetTypeRegistration for () {
    fn get_type_registration2() -> TypeRegistration {
        let mut registration = TypeRegistration::of::<Self>();
        registration.insert::<ReflectResource>(FromType::<Self>::from_type());
        registration
    }
}
```
is what I need to add, however, `()` cannot have `GetTypeRegistration` implemented for it as its outside of the reflect crate, but `ReflectResource` only exists in `bevy_ecs`....

**Comment by alice-i-cecile** [comment]
@MrGVSV @soqb, any advice? We should go ahead and merge this without a `ReflectResource` implementation to start IMO.

**Comment by soqb** [comment]
i'm afraid i don't really understand the motivation. `()` isn't a resource, the implementation in this PR doesn't act like it's a resource, so what's the point?

if something requires this complexity of generics, why not just use `SystemParam`? if there are cases where users want to pass zero resources (instead of one) to an api, wouldn't there also be cases where they want to pass two or more?

even if that doesn't work, it's still better imo to have a trait like `MaybeResource` which is implemented `forall T: Resource` and for `()`, so we don't overcomplicate `Resource`.

**Comment by soqb** [comment]
also, the new `ReflectRef::Unit` doesn't seem to be used anywhere, and so should be removed if/when we merge.

**Comment by Lubba-64** [comment]
> even if that doesn't work, it's still better imo to have a trait like MaybeResource which is implemented forall T: Resource and for (), so we don't overcomplicate Resource.

This seems like a nice idea.

File: crates/bevy_reflect/src/reflect.rs [hunk 1]
```diff
@@ -65,6 +67,7 @@ pub enum ReflectRef<'a> {
     Set(&'a dyn Set),
     Enum(&'a dyn Enum),
     Value(&'a dyn PartialReflect),
+    Unit(&'a ()),
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/reflect.rs:70]: I don't think we should introduce a new `ReflectKind` that's only used by the unit type. I feel it makes more sense to consider it an "empty tuple".

Comment (MrGVSV) [reply crates/bevy_reflect/src/reflect.rs:70]: Actually I think `()` already implements `Tuple`, so I’m not sure this even variant is even used?

File: crates/bevy_ecs/src/world/component_constants.rs [hunk 2]
```diff
@@ -37,3 +39,6 @@ pub struct OnReplace;
 #[derive(Event)]
 #[cfg_attr(feature = "bevy_reflect", derive(Reflect))]
 pub struct OnRemove;
+
+#[cfg(feature = "bevy_reflect")]
```
Comment (MrGVSV) [inline crates/bevy_ecs/src/world/component_constants.rs:43]: Any reason to only allow this impl when `bevy_reflect` is enabled?

**Comment by Lubba-64** [comment]
I am going to close this, I think that there were a lot of flaws with my approach, I think that the idea of making a maybe resource trait is a good idea that I want to try in a new branch.

**Pull request closed without merging**